pylearn2/datasets/hdf5.py

```diff
@@ -2,338 +2,472 @@
-            Order of axes in topological view.
+        sources : list of str
+            List of HDF5 keys corresponding to the data to be loaded.
+        load_all : bool, optional (default False)
+            If true, load dataset into memory.
+        use_h5py: bool, optional (default True)
+            If true uses h5py, else tables.
         """
-        shape = [V.shape[axes.index('b')],
-                 V.shape[axes.index(0)],
-                 V.shape[axes.index(1)],
-                 V.shape[axes.index('c')]]
-        self.view_converter = HDF5ViewConverter(shape[1:], axes=axes)
-        self.X = self.view_converter.topo_view_to_design_mat(V)
-        # self.X_topo_space stores a "default" topological space that
-        # will be used only when self.iterator is called without a
-        # data_specs, and with "topo=True", which is deprecated.
-        self.X_topo_space = self.view_converter.topo_space
-
-        # Update data specs
-        X_space = VectorSpace(dim=V.shape[axes.index('b')])
-        X_source = 'features'
-        if self.y is None:
-            space = X_space
-            source = X_source
+        data = alias_dict()
+        if use_h5py:
+            for s, a in safe_zip(sources, aliases):
+                if load_all:
+                    data[s, a] = self._fhandler[s][:]
+                else:
+                    data[s, a] = self._fhandler[s]
+                    # hdf5 handle has no ndim
+                    data[s].ndim = len(data[s].shape)
         else:
-            if self.y.ndim == 1:
-                dim = 1
-            else:
-                dim = self.y.shape[-1]
-            y_space = VectorSpace(dim=dim)
-            y_source = 'targets'
-            space = CompositeSpace((X_space, y_space))
-            source = (X_source, y_source)
-
-        self.data_specs = (space, source)
-        self.X_space = X_space
-        self._iter_data_specs = (X_space, X_source)
+            for s, a in safe_zip(sources, aliases):
+                if load_all:
+                    data[s, a](self._fhandler.getNode('/', s)[:])
+                else:
+                    data[s, a] = self._fhandler.getNode('/', s)
+        return data
 
+    @wraps(Dataset.iterator, assigned=(), updated=())
+    def iterator(self, mode=None, batch_size=None, num_batches=None,
+                 rng=None, data_specs=None, return_tuple=False,
+                 **kwargs):
+        """
+        data_specs : tuple
+            A `(space, source)` tuple. See :ref:`data_specs` for a full
+            description.
+        return_tuple : bool, optional
+            Always return a tuple, even if there is exactly one source
+            of data being returned. Defaults to `False`.
+        """
+        [mode, batch_size, num_batches, rng, data_specs] = self._init_iterator(
+            mode, batch_size, num_batches, rng, data_specs)
+        convert = None
+
+        if data_specs is None:
+            sources = self._get_canonical_sources()
```

Comment:
Ok, I will add an assert to prevent it to be None. I will also add a comment  both in `HDF5.py` and in `Dataset.py` specifying that we kept it for compatibility with legacy code but it should not be None.